scheduler/executor/workflow_executor.go

```diff
@@ -0,0 +1,349 @@
+package executor
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/flyteorg/flyteadmin/pkg/async/schedule/interfaces"
+	runtimeInterfaces "github.com/flyteorg/flyteadmin/pkg/runtime/interfaces"
+	interfaces2 "github.com/flyteorg/flyteadmin/scheduler/executor/interfaces"
+	"github.com/flyteorg/flyteadmin/scheduler/repositories"
+	"github.com/flyteorg/flyteadmin/scheduler/repositories/models"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/admin"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/core"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+	"github.com/flyteorg/flytestdlib/logger"
+	"github.com/flyteorg/flytestdlib/promutils"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/robfig/cron"
+	"go.uber.org/ratelimit"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"runtime/debug"
+	"strings"
+	"time"
+)
+
+const executorSleepTime = 30
+
+const snapshotWriterSleepTime = 30
+const backOffSleepTime = 60
+const snapShotVersion = 1
+
+type schedulerMetrics struct {
+	Scope                  promutils.Scope
+	FailedExecutionCounter prometheus.Counter
+	CheckPointPanicCounter prometheus.Counter
+	CheckPointSaveErrCounter prometheus.Counter
+	CheckPointCreationErrCounter prometheus.Counter
+	CatchupErrCounter prometheus.Counter
+	ScheduleRegistrationFailure prometheus.Counter
+	ScheduleReadFailure prometheus.Counter
+}
+
+// workflowExecutor used for executing the schedules saved by the native flyte scheduler in the database.
+type workflowExecutor struct {
+	db                   repositories.SchedulerRepoInterface
+	config               runtimeInterfaces.Configuration
+	snapshot             interfaces2.Snapshoter
+	snapShotReaderWriter interfaces2.SnapshotReaderWriter
+	goGfInterface        interfaces2.GoCronWrapper
+	rateLimiter          ratelimit.Limiter
+	metrics              schedulerMetrics
+	adminServiceClient service.AdminServiceClient
+}
+
+func (w *workflowExecutor) CheckPointState(ctx context.Context) {
+	defer func() {
```

Comment:
a better pattern would be put this differ at the call-site. Like in the previous method. then the code here will raise panics and you can test that